src/providers/mod.rs

```diff
@@ -0,0 +1,112 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, you can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::{
+  boxed::Box,
+  collections::HashMap,
+  error::Error,
+  fmt::{self, Display, Formatter},
+};
+
+use self::{mock::MockProvider as Mock, ses::SesProvider as Ses};
+use settings::Settings;
+
+mod mock;
+mod ses;
+
+trait Provider
+{
+  fn send(
+    &self,
+    to: String,
+    cc: Vec<String>,
+    subject: String,
+    body_text: String,
+    body_html: Option<String>,
+  ) -> Result<String, ProviderError>;
+}
+
+#[derive(Debug)]
+pub struct ProviderError
+{
+  description: String,
+}
+
+impl ProviderError
+{
+  pub fn new(description: String) -> ProviderError
+  {
+    ProviderError { description }
+  }
+}
+
+impl Error for ProviderError
+{
+  fn description(&self) -> &str
+  {
+    &self.description
+  }
+}
+
+impl Display for ProviderError
+{
+  fn fmt(&self, formatter: &mut Formatter) -> fmt::Result
+  {
+    write!(formatter, "{}", self.description)
+  }
+}
+
+pub struct Providers
+{
+  settings: Settings,
+  providers: HashMap<String, Box<Provider>>,
+}
+
+impl Providers
+{
+  pub fn new() -> Providers
+  {
+    let mut instance = Providers {
+      settings: Settings::new().expect("config error"),
+      providers: HashMap::new(),
+    };
+    instance
+      .providers
+      .insert(String::from("mock"), Box::new(Mock));
+    instance
+      .providers
+      .insert(String::from("ses"), Box::new(Ses::new(&instance.settings)));
+    instance
+  }
+
+  pub fn send(
+    &self,
+    to: String,
+    cc: Vec<String>,
+    subject: String,
+    body_text: String,
+    body_html: Option<String>,
+    provider_id: Option<String>,
+  ) -> Result<String, ProviderError>
+  {
+    let resolved_provider_id = if let Some(ref id) = provider_id {
+      id
+    } else {
+      &self.settings.provider
+    };
+
+    match self.providers.get(resolved_provider_id) {
+      Some(ref provider) => match provider.send(to, cc, subject, body_text, body_html) {
+        Ok(message_id) => Ok(format!("{}:{}", resolved_provider_id, message_id)),
```

Comment:
Since SES already returns a guaranteed unique message id, for now I'm opting to re-use that and prefix it with the provider id ("ses" in this case). We don't have to keep it that way forever, but at least this lets us postpone the decision about message ids for a while.